icontrol/authtoken.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2015-2016 F5 Networks Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""A requests-compatible system for BIG-IP token-based authentication.
+
+BIG-IP only allows users with the Administrator role to authenticate to
+iControl using HTTP Basic auth.  Non-Administrator users can use the
+token-based authentication scheme described at:
+
+  https://devcentral.f5.com/wiki/icontrol.authentication_with_the_f5_rest_api.ashx
+
+Use this module with requests to automatically get a new token, and attach
+:class:`requests.Session` object, so that it is used to authenticate future
+requests.
+
+This can be enabled in the iControlRESTSession by passing a "token=True"
+argument:
+
+   iCRSession = iControlRESTSession('bob', 'secret', token=True)
+"""
+
+from icontrol.exceptions import iControlUnexpectedHTTPError
+from icontrol.exceptions import InvalidScheme
+import requests
+from requests.auth import AuthBase
+from requests.auth import HTTPBasicAuth
+import time
+import urlparse
+
+
+class iControlRESTTokenAuth(AuthBase):
+    def __init__(self, username, password, login_provider_name='tmos'):
+        self.username = username
+        self.password = password
+        self.login_provider_name = login_provider_name
+        self.token = None
+        self.expiration = None
+        self.attempts = 0
+        # We don't actually do auth at this point because we don't have a
+        # hostname to authenticate to.
+
+    def token_valid(self):
```

Comment:
I like methods to be verby, also this method appears to only be called internally, so I propose we change its name to something like:

`_check_token_validity`

I don't really care about the exact name as long as:

It starts with an `_`, and ideally is a verb.   I care more about the `_` part.